### PR TITLE
Introduce CoroutineDispatcher.asExecutor() extension

### LINKED ISSUE #1450: Convert `CoroutineContext` into `Executor`
I have to use an API that accepts an `Executor` as a parameter. How can I create an `Executor` that is backed by `Dispatchers.Default`? 

Can a `CoroutineDispatcher.asExecutor(): Executor` function be added to support this use case?

```kotlin
fun externalApi(executor: Executor)

fun example() {
  val executor: Executor = Dispatchers.Default.asExecutor()
  
  externalApi(executor)
}
```

Even better would be a `fun CoroutineContext.asExecutor(): Executor` function. I would imagine if the receiver `CoroutineContext` does not have a `ContinuationInterceptor`, this function would throw.

### PR DESCRIPTION
**Pull request by elizarov**
Fixes #1450

### CODE REVIEW

File: kotlinx-coroutines-core/jvm/test/ExecutorsTest.kt [hunk 1]
```diff
@@ -32,14 +33,32 @@ class ExecutorsTest : TestBase() {
...
+    }
+
+    @Test
+    fun testDefaultDispatcherToExecutor() {
+        val latch = CountDownLatch(2)
+        Dispatchers.Default.asExecutor().execute {
+            checkThreadName("DefaultDispatcher")
+            latch.countDown()
+        }
+        latch.countDown()
```
Comment (JakeWharton) [inline kotlinx-coroutines-core/jvm/test/ExecutorsTest.kt]: await?

File: kotlinx-coroutines-core/jvm/src/Executors.kt [hunk 2]
```diff
@@ -45,6 +45,19 @@ public fun ExecutorService.asCoroutineDispatcher(): ExecutorCoroutineDispatcher
 public fun Executor.asCoroutineDispatcher(): CoroutineDispatcher =
     ExecutorCoroutineDispatcherImpl(this)
 
+/**
+ * Converts an instance of [CoroutineDispatcher] to an implementation of [Executor].
+ *
+ * It returns an original executor when used on the result of [Executor.asCoroutineDispatcher] extensions.
```
Comment (paulblessing) [inline kotlinx-coroutines-core/jvm/src/Executors.kt]:
```suggestion
 * It returns the original executor when used on the result of [Executor.asCoroutineDispatcher] extensions.
```

File: kotlinx-coroutines-core/jvm/test/ExecutorsTest.kt (hunk 1 above)
Comment (elizarov) [reply kotlinx-coroutines-core/jvm/test/ExecutorsTest.kt]: Good catch.

File: kotlinx-coroutines-core/jvm/test/ExecutorsTest.kt [hunk 3]
```diff
@@ -32,14 +34,49 @@ class ExecutorsTest : TestBase() {
...
         }
         executor.shutdown()
     }
 
+    @Test
+    fun testConvertedDispatcherToExecutor() {
+        val executor: ExecutorService = Executors.newSingleThreadExecutor { r -> Thread(r, "TestExecutor") }
+        val dispatcher: CoroutineDispatcher = executor.asCoroutineDispatcher()
+        assertSame(executor, dispatcher.asExecutor())
+        executor.shutdown()
```
Comment (qwwdfsad) [inline kotlinx-coroutines-core/jvm/test/ExecutorsTest.kt]: Note: `shutdown` doesn't await actual shutdown

Comment (elizarov) [reply kotlinx-coroutines-core/jvm/test/ExecutorsTest.kt]: That's Ok. Infrastrusture in `TestBase` awaits.

Comment (JakeWharton) [reply kotlinx-coroutines-core/jvm/test/ExecutorsTest.kt]: It looks like it only awaits DefaultExecutor (and other dispatchers). You would need to register this out-of-band `Executor` with it somehow or put a factory method in `TestBase`.

Comment (elizarov) [reply kotlinx-coroutines-core/jvm/test/ExecutorsTest.kt]: https://github.com/Kotlin/kotlinx.coroutines/blob/cd5e60e9f2470dced9687cd908b13c9c9aef887d/kotlinx-coroutines-core/jvm/test/Threads.kt#L47-L53

Comment (JakeWharton) [reply kotlinx-coroutines-core/jvm/test/ExecutorsTest.kt]: Clever!

**Pull request merged**